_posts/2015-03-23-design-patterns-part-1.markdown

```diff
@@ -0,0 +1,131 @@
+---
+
+<meta charset="utf-8">
+<!-- not sure this charset isn't pulled in from _includes --> 
+
+# Design Patterns, Fast Tests, and Robust Software, Part I
+
+Writing high quality software means expecting change: changes to business requirements, changes to the web framework you’ve chosen, changes to the
+libraries you depend on and changes to the coding language.  At SitterCity, we typically rely on Rails for our web framework, but our Rails projects don't looks like most Rails projects.  When we're building a Rails app, we try to leverage the good parts of Rails and avoid the bad.  We typically construct views and controllers in a conventional way, but handle persistent storage much differently than an idiomatic rails app.
+
+Part of anticipating change means that we structure our software in a way that minimizes its reliance on the web framework.  To accomplish this, we
+rely on a constellation of design patterns that includes repositories, entities, dependency injection, factories, and commands (sometimes called
+contexts or use-cases).  I won't spend a lot of time explaining each pattern, so if these are new to you, then I'd suggest you look at the relevant
+chapters of <a href="http://designpatternsinruby.com"> this book </a>.  Most of what we sketch out in these posts will be a rough approximation of
+what you would write in a real project.  I’m omitting details about namespacing, error handling, and testing nuances so that we can focus on the big
+picture. In this series of blog posts, I’ll explain how these patterns work in concert to provide us an app that is responsible, resilient, and robust.
+
+In order to explain this architecture, let's use the example of creating a user..  
+
+## Repositories and Entities
+
+Most Rails apps use the ORM that ships with Rails, ActiveRecord.  To create a user, we'd have something that looks like this.
+
+{% highlight ruby %}
+
+    User.create(first_name: "John", last_name: "Doe", email: "JohnDoe@Example.com")
+
+{% endhighlight %}
+
+At SitterCity, we often want the flexibility to change parts of the app, and that includes changing the ORM.  So, we'll use the repository pattern concentrate all ActiveRecord queries into single class.  Instead of the code you saw above, we might have a class that looks like this:
+
+
+{% highlight ruby %}
+
+class ARUserRepository
+
+  def create(first_name, last_name, email)
+    User.create(first_name: first_name, last_name: last_name, email)
+  end
+
+end
+
+{% endhighlight %}
+
+If we decide that we don't want to use ActiveRecord anymore, there won't be a lot of work involved.  Let's say that we want to use the Sequel ORM.  We'd end up writing another repository that had the exact same interface as the ARUserRepository.  (If you’re unfamiliar with the idea of interfaces, just know that in the world of Ruby, we can say that two classes have the same interface if they have the same public methods available to them.)
```

Comment:
I like this but maybe we can add a link to something about duck typing? We don't have to go into it but at least mentioning duck typing is a good thing here, I think.